### PR TITLE
[WIP] Update cuDF UDF Notebook to include apply_grouped

### PR DESCRIPTION
**Pull request by beckernick**
Lorem ipsum.

### CODE REVIEW

**Comment by scopatz** [comment]
Does this need any help etc @beckernick?

**Comment by beckernick** [comment]
~This should be largely good to go @scopatz. I'll review it again tonight (pacific time), but if it looks good to you feel free to merge~.

This need some minor edits, which I will get done tonight PST

File: cudf/cuDF UDF.ipynb [hunk 1]
```diff
@@ -4,17 +4,16 @@
...
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Sometimes, the builtin methods of cudf.DataFrame don't do exactly what we want. We need to write a custom function to apply over the data frame.\n",
+    "Sometimes, the builtin methods of cudf.DataFrame don't do exactly what we want. We need to write a custom function (also known as a user defined function) to apply over the DataFrame.\n",
     "\n",
-    "cuDF’s DataFrame class has two methods that let users run custom Python functions on GPUs: `apply_rows` and `apply_chunks`. In this tutorial, we’ll walk through how to use `apply_rows` and `apply_chunks` to create your own UDFs and show how you can implement a GPU-accelerated windowing function.\n",
-    "\n"
+    "cuDF’s DataFrame class has two primary methods that let users run custom Python functions on GPUs: `apply_rows` and `apply_chunks`. In this tutorial, we’ll walk through how to use `apply_rows` and `apply_chunks` to create your own UDFs and show how you can implement a GPU-accelerated windowing function. At the end, we'll also walk through a more advanced example of applying a user defined function on a grouped DataFrame (using `apply_grouped`).\n"
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb:16]: No longer doing window function example

File: cudf/cuDF UDF.ipynb [hunk 2]
```diff
@@ -30,7 +29,16 @@
...
+    "\n",
+    "How does this work? User defined functions with cuDF rely on CUDA under the hood. Exploring CUDA in-depth is out of scope for this tutorial. But, at a very high level, in cuDF's user defined functions:\n",
+    "\n",
+    "\n",
+    "- Compute is spread across multiple \"blocks\", which have access to both global memory but also their own in-block memory \n",
+    "- Within each block, many \"threads\" operate independently and can quickly access data in their block-specific shared memory \n",
+    "\n",
+    "\n",
+    "As a result, the loop in the example function below resembles serial code, but executes concurrently in multiple threads on the GPU. When `kernel` is invoked, the function arguments corresponding to the input/output are strided so as to improve GPU parallelism. The kernel function is compiled to the GPU using `numba.cuda`, so the kernel function must only use Python features/functions that are [supported](https://numba.pydata.org/numba-doc/dev/cuda/cudapysupported.html) by Numba for CUDA."
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb]: executes in parallel

File: cudf/cuDF UDF.ipynb [hunk 3]
```diff
@@ -46,7 +54,7 @@
     "        out2[i] = ...\n",
     "```\n",
     "\n",
-    "`in1, in2, in3, ...` are the input columns. `out1, out2, ...` are the output columns. The kernel function should not return a result. Rather, output columns are passed as arguments and the result is written to them. Additional keyword arguments can be passed (`kwarg1, kwarg2, ...`). Inside the kernel function, [standard numba.cuda attributes](https://numba.pydata.org/numba-doc/dev/cuda/kernels.html#thread-positioning) like `numba.cuda.threadIdx` can be used to access things like the thread or block indices.\n",
+    "`in1, in2, in3, ...` are the input columns. `out1, out2, ...` are the output columns. The kernel function should not return a result. Rather, output columns are passed as arguments and the result is written to them. Additional keyword arguments can be passed (`kwarg1, kwarg2, ...`). Inside the kernel function, [standard numba.cuda attributes](https://numba.pydata.org/numba-doc/dev/cuda/kernels.html#thread-positioning) like `numba.cuda.threadIdx` can be used to access things like the thread or block indices. We'll explain a little bit more about this below.\n",
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb]: explain the indexing of the out array

File: cudf/cuDF UDF.ipynb [hunk 4]
```diff
@@ -46,7 +54,7 @@
...
     "\n",
     "`apply_rows` is called like\n",
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb]: Add a transition

File: cudf/cuDF UDF.ipynb [hunk 5]
```diff
@@ -204,7 +213,13 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "In the example above, the sample output in the last cell shows that the for-loop in the kernel function is unrolled by the compiler automatically. It uses 15 CUDA blocks, each using 64 threads to do the computation. In this case, most threads in a block handle one element from the input array, but some threads have to deal with two elements, because there are 1000 rows and 960 threads (15 blocks * 64 threads per block). Additionally, due to the parallelism, the DataFrame rows are not necessarily processed in order."
+    "In the example above, the printed output from running our applying our `haversine_distance_kernel` function shows some pretty informative information. If you were to look at the entire printed output, you'd notice a few things:\n",
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb]: grammar (running applying)

File: cudf/cuDF UDF.ipynb [hunk 6]
```diff
@@ -315,16 +330,18 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "`apply_chunks` is a more general version of `apply_rows`, that gives us control over how the data is chunked on the GPU. We can specify how to divide the long array into chunks, map each of the array chunks to different GPU blocks to process (chunks argument) and assign the number of threads per block (tpb argument). The for-loop is also no longer automatically unrolled in the kernel function, instead serving as the for-loop for that GPU thread.\n",
+    "In the section above, we the data was generally split into single-element chunks. `apply_chunks` is a more general version of `apply_rows` that gives us control over how the data is chunked on the GPU. We can specify how to divide the long array, map each of the array chunks to different GPU blocks to process (using the `chunks` argument) and assign the number of threads per block (using the `tpb` argument).\n",
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb]: grammar (we/the)

File: cudf/cuDF UDF.ipynb [hunk 7]
```diff
@@ -315,16 +330,18 @@
...
     "\n",
-    "`apply_chunks` kernels work the same as as `apply_rows` kernels. When we call `apply_chunks`, we must supply the chunk size `chunks` as an integer or `cudf.Series` of integer offsets, and the number of threads per block `tpb`. Note that `tpb` can be omitted, but in that case, it defaults to `1` thread per block, which is very inefficient. Therefore it is recommended to always set this argument.\n",
+    "Applying kernels with `apply_chunks` is very similar to applying kernels with `apply_rows`. Except, when we call `apply_chunks`, we must also provide:\n",
+    "- The chunk size `chunks` as an integer or `cudf.Series` of integer offsets\n",
+    "- The number of threads per blockm `tpb`. Note that `tpb` can be omitted, but in that case, it defaults to `1` thread per block, which is very inefficient. We recommend always setting this argument.\n",
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb]: blockm

File: cudf/cuDF UDF.ipynb [hunk 8]
```diff
@@ -429,149 +425,169 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Advanced `apply_chunks`: rolling window"
+    "# Advanced UDFs: `apply_grouped`"
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb:436]: Transition. Mention that there is actually another way to do UDFs

File: cudf/cuDF UDF.ipynb (hunk 1 above)
Comment (beckernick) [reply cudf/cuDF UDF.ipynb:16]: Nevermind. We are

File: cudf/cuDF UDF.ipynb [hunk 9]
```diff
@@ -429,149 +425,169 @@
...
-      "7  7.0  5.5\n",
-      "8  8.0  6.5\n",
-      "9  9.0  7.5\n"
-     ]
-    }
-   ],
    "source": [
-    "print(df.head(10))"
+    "Notice that the value for `avg` at beginning of each group is null. This makes sense, since we can't have an average value for a sliding window of size `3` until we have at least three values. We used the line `avg[i] = np.nan` in our `rolling_avg` function to geneate this result in the output."
    ]
```
Comment (beckernick) [inline cudf/cuDF UDF.ipynb:550]: Add an outro sentence

**Comment by beckernick** [comment]
Updated notebook. Merging.

**Pull request merged**